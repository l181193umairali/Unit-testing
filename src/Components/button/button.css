.button-style
{
    border: 1px solid yellow;
    padding: 10px;
}
.button-style3
{
    border: 4px black;
    padding: 40px;
}