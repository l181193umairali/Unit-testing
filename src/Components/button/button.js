import React from "react";
import './button.css'
export default function Button({label})
{
    return <div data-testid="button"className="button-style">{label}</div>
    //data is the special attribute used in html 
}
export function Button2({label})
{
    return <div data-testid="button1"className="button-style3">{label}</div>
}


