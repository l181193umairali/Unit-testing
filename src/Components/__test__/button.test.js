import React from 'react'
// import ReactDOM from 'react-dom'; 
import Button from '../button/button';
import { Button2  } from '../button/button';
import "@testing-library/jest-dom"
import { createRoot } from 'react-dom/client';
// ReactDOM.render is no longer supported in React 18. Use createRoot instead. 
// Until you switch to the new API, your app will behave as if it’s running React 17.
import {  render ,screen,cleanup} from '@testing-library/react';
// import {TestEvents,Button} from '../button/button'
import ShallowRenderer from 'react-shallow-renderer';

// cleanup is passed as a parameter to afterEach to just clean up everything after each test to avoid memory leaks.

afterEach(cleanup)

it("rendres without crashing",()=>
{
    const div=document.createElement("div");
    const root = createRoot(div);
    // we are gonna attach that component with the div
    root.render(<Button></Button>,div);
});
it("render button correctly",()=>
{
    render(<Button label="click"/>)
    expect(screen.getByTestId('button')).toBeInTheDocument("click")
});

it("render button2 correctly",()=>
{
    render(<Button2 label="save"/>)
    expect(screen.getByTestId('button1')).toBeInTheDocument("save")
});
// When you're writing tests, you often need to check that values meet certain conditions. 
// expect gives you access to a number of "matchers" that let you validate different things.
it("matches snapshots",()=>
{
    render(<Button label="click"/>)
    expect(screen.getByTestId('button')).toMatchSnapshot("click")
    
});
// How snapshot testing works is that it compares the component with 
// the generated blueprint (snapshot). If they don't match, the test will fail.
it("matches snapshots 2",()=>
{
    render(<Button label="click"/>)
    expect(screen.getByTestId('button')).toMatchSnapshot("click me please")
    
});

//Sallow rendering
// You can think of the shallowRenderer as a “place” to render the component
//  you’re testing, and from which you can extract the component’s output.

// shallowRenderer.render()
//  is similar to root.render() but it doesn’t require
//   DOM and only renders a single level deep. 
//   This means you can test components isolated from how their children are implemented.

// const renderer = new ShallowRenderer();
// render(<span className="heading"/>)
// renderer.getRenderOutput();
// expect(renderer.type).toBe('div');
// expect(renderer.props).toEqual([
//   <span className="heading">Title</span>
// ]);