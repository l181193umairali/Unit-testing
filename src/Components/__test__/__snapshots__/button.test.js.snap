// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 2: click me please 1`] = `
<div
  class="button-style"
  data-testid="button"
>
  click
</div>
`;

exports[`matches snapshots: click 1`] = `
<div
  class="button-style"
  data-testid="button"
>
  click
</div>
`;
